lib/server-builder.js

```diff
@@ -0,0 +1,161 @@
+import { SubProcess } from 'teen_process';
+import { fs, logger, system } from 'appium-support';
+import log from './logger';
+import path from 'path';
+import { EOL } from 'os';
+
+const IGNORED_ERRORS = [];
+const GRADLE_VERSION_KEY = 'gradle';
+const GRADLE_URL_PREFIX = 'distributionUrl=';
+const GRADLE_URL_TEMPLATE = 'https\\://services.gradle.org/distributions/gradle-VERSION-all.zip';
+const GRADLE_MAX_ERROR_LOG_LINES = 15;
+
+const GRADLE_DEPENDENCIES_PLACEHOLDER = '// additionalAppDependencies placeholder (don\'t change or delete this line)';
+
+const VERSION_KEYS = [
+  GRADLE_VERSION_KEY,
+  'androidGradlePlugin',
+  'compileSdk',
+  'buildTools',
+  'minSdk',
+  'targetSdk',
+  'kotlin'
+];
+
+const gradleLog = logger.getLogger('Gradle');
+
+class ServerBuilder {
+  constructor (args = {}) {
+    this.serverPath = args.serverPath;
+    this.showGradleLog = args.showGradleLog;
+
+    const buildConfiguration = args.buildConfiguration || {};
+
+    const versionConfiguration = buildConfiguration.toolsVersions || {};
+    const serverVersions = {};
+    for (const key of VERSION_KEYS) {
+      if (versionConfiguration[key]) {
+        serverVersions[key] = versionConfiguration[key];
+      }
+    }
+    this.serverVersions = serverVersions;
+
+    this.additionalAppDependencies = buildConfiguration.additionalAppDependencies || [];
+  }
+
+  async build () {
+    if (this.serverVersions[GRADLE_VERSION_KEY]) {
+      await this.setGradleWrapperVersion(this.serverVersions[GRADLE_VERSION_KEY]);
+    }
+    await this.insertAdditionalAppDependencies(this.additionalAppDependencies);
```

Comment:
should we check if this.additionalAppDependencies is not empty first?

Reply:
It is not required as current implementation should not break anything. But yes, that will remove some IO calls when they are not required.